apps/store/src/services/shopSession/ShopSessionContext.tsx

```diff
@@ -46,55 +35,60 @@ export const ShopSessionProvider = ({ children }: PropsWithChildren<unknown>) =>
       }
     },
     onError: (error) => {
-      console.warn('ShopSession not found: ', shopSessionId)
-      console.warn(error)
+      console.warn('ShopSession not found: ', shopSessionId, error)
       createShopSession()
     },
   })
 
-  const value = useMemo<ShopSessionContextValue>(() => {
-    const shopSessionCountryCode = result.data?.shopSession?.countryCode
-    return {
-      shopSession: shopSessionCountryCode !== countryCode ? null : result.data?.shopSession ?? null,
-      loading: result.loading,
+  // const shopSessionCountryCode = result.data?.shopSession?.countryCode
+  // FIXME: Ensure we don't return incorrect country data.  Perhaps remove session from apolloCache
+  // shopSession: shopSessionCountryCode !== countryCode ? null : result.data?.shopSession ?? null,
+
+  // Has to be wrapped to prevent duplicate execution (Apollo)
+  useEffect(() => {
+    // TODO: isBrowser() and ensure code splitting does not break
+    if (typeof window !== 'undefined' && !shopSessionId && !mutationResult.called) {
+      createShopSession()
     }
-  }, [result, countryCode])
+  }, [createShopSession, mutationResult.called, shopSessionId])
 
-  return <ShopSessionContext.Provider value={value}>{children}</ShopSessionContext.Provider>
+  return <ShopSessionContext.Provider value={queryResult}>{children}</ShopSessionContext.Provider>
 }
 
 export const useShopSession = () => {
-  return useContext(ShopSessionContext)
+  const queryResult = useContext(ShopSessionContext)
+  if (!queryResult) {
+    throw new Error(
+      'useShopSession called from outside ShopSessionContextProvider, no value in context',
+    )
+  }
+  return queryResult
```

Comment:
Suggestion:

```suggestion
  return queryResult?.data?.shopSession
```

I think it would make the hook more _ergonometric_ by not having to do `data.shopSession` every place we use it.

Reply:
We want to keep an option to use other properties of returned object, like `refetch`, `loading`, etc

Do you think it makes sense to destructure the data and keep everything else the same?  What worries me is the amount of type declarations to maintain to keep this code type-safe

Reply:
Hmm I think we can just use `ShopSessionQuery['shopSession']`

Perhaps that would be worth it.

Reply:
I see. But since you've mentioned the increasing of type declarations, was you considering into create a new type that defines the return of `useShopSession`? Or do I misunderstood? 🤔

Reply:
We could do something like this
```
  const { data, ...other } = useShopSessionApolloQuery({
    variables: shopSessionId ? { shopSessionId } : undefined,
    skip: !shopSessionId,
    // Only intended to run client-side, prefetch and pass to Apollo Cache if fetched on server.
    ssr: false,
    ...rest,
  })
  return {
    ...other,
    shopSession: data?.shopSession,
  }
```

It actually infers return type, no need to declare anything

What do you think?  @guilhermespopolin @robinandeer

Reply:
This requires us to make context value type more complex though
```
type ShopSessionQueryResult = Omit<
  QueryResult<ShopSessionQuery, ShopSessionQueryVariables>,
  'data'
> & { shopSession: ShopSessionQuery['shopSession'] }
```

Reply:
Not too bad, we could give it a go -- but you can also do that in another PR if you like!

Reply:
Oh, I see. I guess we could keep `ShopSessionQueryResult` as we had before and make the transformation on `useShopSession`

```javascript
type ShopSessionQueryResult = QueryResult<ShopSessionQuery, ShopSessionQueryVariables>

...

export const useShopSession = () => {
  const queryResult = useContext(ShopSessionContext)

  if (!queryResult) {
    throw new Error(
      'useShopSession called from outside ShopSessionContextProvider, no value in context',
    )
  }

  const { data, ...other } = queryResult
  return {
    shopSession: data?.shopSession ?? null,
    ...other,
  }
}

...

const { shopSession } = useShopSession()
```

In that way we don't need to create complex type for the provider and the return type of `useShopSession` can be inferred. But either way is fine by me 😄

Reply:
Sounds great, I'll move transform logic to useShopSession.  Makes each of the hooks cleaner